geoana/em/static/halfspace.py

```diff
@@ -0,0 +1,628 @@
+        """
+
+        xyz = check_xyz_dim(xyz)
+        if np.any(xyz[..., -1] > 0):
+            raise ValueError(
+                f"z value must be less than or equal to 0 in a halfspace, got {(xyz[..., -1])}"
+            )
+
+        e = self._primary.electric_field(xyz) + self._image.electric_field(xyz)
+        return e
+
+    def current_density(self, xyz):
+        """Current density for a point current in a halfspace.
+
+       This method computes the current density for the point current in a halfspace at
+        the set of gridded xyz locations provided. Where :math:`\\rho`
+        is the electric resistivity and :math:`\\mathbf{E}` is the electric field
+        for the point current.
+        The current density :math:`\\mathbf{J}` is:
+
+       .. math::
+
+            \\mathbf{J} = \\frac{\\mathbf{E}}{\\rho}
+
+        Parameters
+        ----------
+        xyz : (..., 3) numpy.ndarray
+            Locations to evaluate at in units m.
+
+        Returns
+        -------
+        J : (..., 3) np.ndarray
+            Current density of point current in units :math:`\\frac{A}{m^2}`.
+
+        Examples
+        --------
+        Here, we define a point current with current=1A in a halfspace and plot the current density.
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> from geoana.em.static import PointCurrentHalfSpace
+
+        Define the point current.
+
+        >>> rho = 1.0
+        >>> current = 1.0
+        >>> simulation = PointCurrentHalfSpace(
+        >>>     current=current, rho=rho, location=None
+        >>> )
+
+        Now we create a set of gridded locations and compute the current density.
+
+        >>> X, Y = np.meshgrid(np.linspace(-1, 1, 20), np.linspace(-1, 1, 20))
+        >>> Z = np.zeros_like(X)
+        >>> xyz = np.stack((X, Y, Z), axis=-1)
+        >>> j = simulation.current_density(xyz)
+
+        Finally, we plot the curent density.
+
+        >>> j_amp = np.linalg.norm(j, axis=-1)
+        >>> plt.pcolor(X, Y, j_amp, shading='auto')
+        >>> cb1 = plt.colorbar()
+        >>> cb1.set_label(label= 'Current Density ($A/m^2$)')
+        >>> plt.ylabel('Y coordinate ($m$)')
+        >>> plt.xlabel('X coordinate ($m$)')
+        >>> plt.title('Current Density for a Point Current in a Halfspace')
+        >>> plt.show()
+        """
+
+        j = self.electric_field(xyz) / self.rho
+        return j
+
+
+class DipoleHalfSpace:
+    """Class for a dipole in a halfspace.
+
+        The ``DipoleHalfSpace`` class is used to analytically compute the
+        potentials, current densities and electric fields within a halfspace due to a dipole source.
+        Surface is assumed to be at z=0.
+
+        Parameters
+        ----------
+        current : float
+            Electrical current in the point current (A). Default is 1A.
+        rho : float
+            Resistivity in the point current (:math:`\\Omega \\cdot m`).
+        locations : array_like
```

Comment:
For this Dipole, imagine there are two PointCurrentHalfSpace objects in this class, one at a `location_A` and another at a `location_B`. It's probably best to have those be two separate inputs to the `__init__` (and two separate properties of the class) for the most clarity. These would then be used to create two of the `PointCurrentHalfSpace` objects, one at the `A` electrode, and another at the `B` electrode. So instead of `_primary` and `_secondary` for clarity I'd name them something like `_a` and `_b`, or at least something referring to those names.

For defaults, you could always do, `[-1, 0, 0]` and `[1, 0, 0]`, whereas for the pole your default was `[0, 0, 0]`